contracts/VotingRewards.sol

```diff
@@ -0,0 +1,502 @@
+    }
+
+    /**
+     * @notice collect rewards for an address if lockTime is passed since when tokens have been distributed
+     * @param _beneficiary address that should be fund with rewards
+     * @dev rewardsVault should have TRANSFER_ROLE permission
+     */
+    function collectRewardsFor(address _beneficiary) public {
+        uint64 timestamp = getBlockNumber64();
+        // prettier-ignore
+        Reward[] storage rewards = addressRewards[_beneficiary];
+
+        require(rewards.length > 0, ERROR_NO_REWARDS);
+
+        uint256 collectedRewards = 0;
+        for (uint256 i = 0; i < rewards.length; i++) {
+            if (
+                timestamp - rewards[i].lockBlock > rewards[i].lockTime &&
+                rewards[i].state != RewardState.Withdrawn
+            ) {
+                rewards[i].state = RewardState.Withdrawn;
+                rewardsVault.transfer(
+                    rewardsToken,
+                    _beneficiary,
+                    rewards[i].amount
+                );
+                collectedRewards = collectedRewards.add(1);
+
+                emit RewardCollected(_beneficiary, rewards[i].amount);
+            }
+        }
+
+        require(collectedRewards >= 1, ERROR_NO_REWARDS);
+    }
+
+    /**
+     * @notice Reward is calculated as the minimum balance between the
+     *         end of an epoch (now) and the balance at the first
+     *         vote in an epoch (in percentage) for each vote happened within the epoch
+     * @param _beneficiary beneficiary
+     * @param _fromBlock block from wich starting looking for votes
+     * @param _toBlock   block to wich stopping looking for votes
+     * @param _missingVotesThreshold number of vote to which is possible to don't vote
+     */
+    function _calculateReward(
+        address _beneficiary,
+        uint64 _fromBlock,
+        uint64 _toBlock,
+        uint256 _missingVotesThreshold
+    ) internal view returns (uint256) {
+        uint256 missingVotes = 0;
+        uint256 minimunBalance = 0;
+        bool forceFirstAssignment = true;
+
+        // voteId starts from 1 in DandelionVoting
+        for (
+            uint256 voteId = dandelionVoting.votesLength() + 1;
+            voteId > 1;
+            voteId--
+        ) {
+            uint256 realVoteId = voteId.sub(1);
+            uint64 startBlock;
+            (, , startBlock, , , , , , , , ) = dandelionVoting.getVote(
+                realVoteId
+            );
+
+            if (startBlock >= _fromBlock && startBlock <= _toBlock) {
+                DandelionVoting.VoterState state = dandelionVoting
+                    .getVoterState(realVoteId, _beneficiary);
+
+                if (state == DandelionVoting.VoterState.Absent) {
+                    missingVotes = missingVotes.add(1);
+                }
+
+                require(
+                    missingVotes <= _missingVotesThreshold,
+                    ERROR_TOO_MANY_MISSING_VOTES
+                );
```

Comment:
This worries me - having this `require` in here proves that this function is not tested enough, as if it was one of your tests would ought now to be failing.

If a user did not miss voting on the first vote checked on the first turn of this loop, `missingVotes` here will **always** be 0 and thus **always** be failing.  

In fact, if the `_missingVoteThreshold` is anything > 1, we can _never_ pass this require statement.